src/app/icd/client/CheckInDelegate.h

```diff
@@ -39,6 +42,38 @@ class DLL_EXPORT CheckInDelegate
                                node that sent the check-in message.
      */
     virtual void OnCheckInComplete(const ICDClientInfo & clientInfo) = 0;
+
+    /**
+     * @brief Callback used to let the application know that a key refresh is
+     * needed to avoid counter rollover problems.
+     *
+     * The implementer of this function should create a new RefreshKeySender object. They should also generate a new key and store
+       it in the RefreshKeySender object. This object will be tied to the specific key refresh process and will not be used by the
+       caller after that particular key refresh process has ended, regardless of success or failure.
+
+     * If the callee is unable to provide the RefreshKeySender object, that indicates key
+     * refresh is not possible until the callee is able to provide the required resources.
+     *
+     * @param[in] clientInfo - ICDClientInfo object representing the state associated with the
+                               node that sent the check-in message. The callee can use the clientInfo to determine the type of key
+                               to generate.
+     * @param[in] clientStorage - ICDClientStorage object stores the updated ICDClientInfo after re-registration into
+                                  persistent storage.
+     * @return RefreshKeySender - pointer to RefreshKeySender object
+     */
+    virtual RefreshKeySender * OnKeyRefreshNeeded(ICDClientInfo & clientInfo, ICDClientStorage * clientStorage) = 0;
+
+    /**
+ * @brief Callback used to let the application know that the re-registration process is done. This callback will be called for both
+          success and failure cases. On failure, the callee should address the error reason and take appropriate corrective action.
+ *
+ * @param[in] clientInfo - ICDClientInfo object representing the state associated with the
+                           node that sent the check-in message. This will have the new key used for registration and the updated
+                           icd counter.
+   @param[in] aError - CHIP_NO_ERROR indicates successful re-registration using the new key
+                       Other errors indicate the failure reason.
+ */
+    virtual void OnKeyRefreshDone(const ICDClientInfo & clientInfo, CHIP_ERROR aError) = 0;
```

Comment:
OnKeyRefreshDone may just take RefreshKeySender, then application would clearly know the target we need delete.
We don't need to expose clientInfo again to application, OnCheckInComplete has done that, right?

Reply:
Yes, that makes sense. Modified it.